Deprecate camel case attributes

jmks opened the pull request:
# Pull Request

## Related issue
Fixes #320 

## What does this PR do?

When calling `index#search` with non-snake case attributes (e.g. `camelCase` or `weirdMixed_Case`) a warning is emitted (to stderr). Here's an example:

```
Attributes will be expected to be snake_case in future versions of MeiliSearch.

Non-conforming attributes: attributesToHighlight
```
I have a few questions:

1. Which attributes?

The issue mentioned `sending the search attributes`. Do you want this deprecation warning for just index#search or all methods (i.e. any caller of `Utils.transform_attributes`)?

Currently this just handles index#search but should be easy adapt for all callers.

2. The deprecation warning message

I wrote something I would like to see but I'm open to suggestions.

3. `Utils.transform_attributes` handles an Array argument.

I ran the test suite and did not find an example of an array passed to this function. Do you know when this could be called with an Array argument?

4. I didn't see any examples in `.code-samples.meilisearch.yml` that needed changing (again, only for index#search).

However, I did update the specs.

Cheers!

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

brunoocasali commented on lib/meilisearch/utils.rb:
```diff
@@ -23,6 +23,20 @@ def self.parse(body)
+      return if non_snake_case.empty?
+
+      message = <<~MSG
+        Attributes will be expected to be snake_case in future versions of MeiliSearch.
```
```suggestion
        Attributes will be expected to be snake_case in future versions of Meilisearch Ruby.
```

brunoocasali commented:
Hi, @jmks thank you so much for your contribution! It is amazing! 

Let's answer your questions:

> Which attributes?

I think we can use the same logic for every method that has a `transform_attributes` call because the `transform_attributes` is the one responsible for dealing with the possibility of handling snake/camel casing.


> The deprecation warning message

I just sent a recommendation because of the Meilisearch name, but it is ok for me.

> Utils.transform_attributes handles an Array argument.

Yeah, I didn't have much context of the code when I wrote it in the first place, but today I don't know any use case for that handler. But to be honest this code will be removed in a couple of releases after the next one, because now we will have the deprecation warning :)

brunoocasali reviewed (approved):
bors merge

brunoocasali commented:
bors cancel

brunoocasali commented:
_This message is sent automatically_

Thanks again for contributing to Meilisearch :heart:
If you are participating in [Hacktoberfest](https://blog.meilisearch.com/contribute-hacktoberfest-2022/), and you would like to receive some gift from Meilisearch too, please complete [this form](https://meilisearch.typeform.com/to/idt5qugY).

brunoocasali commented:
Hi, @jmks do you still have time to finish this PR? Otherwise I can do it anyway :)

jmks commented:
Sorry, didn't mean to sit on this for so long. I did start working on it again today and will update the PR in the next day or so.

jmks commented:
Hi!

I've updated the all callers of `transform_attributes` to emit warnings now and updated all the specs to not emit warnings.

> But to be honest this code will be removed in a couple of releases after the next one, because now we will have the deprecation warning :)

I'm not sure about that. My impression of this was that in the Ruby code, this is standardizing on snake_case (as per Ruby idiom) but when sending to the Meilisearch server (or receiving responses), the attributes will still be camelCase.
